### PR TITLE
Closes #8004 add param url and refresh

### LINKED ISSUE #8004: Add a parameter in dashboard_url based on it it will refetch user data after rocketcdn express checkout purchase is completed
We want to add a new parameter in `dashboard_url` as we did in Rocket Insights `rocket_insights_upgrade`.
Based on the new rocketcdn parameter
- The plugin will (clear a transient) refetch user data from `/stat/1.0/wp-rocket/user.php`, and it will return
`cdn_token`, `cdn_url` and `rocketcdn_website_id`
- Do the necessary configuration of RocketCDN on the customer's website `wp_ajax_rocketcdn_validate_token_cname` (AI plugin uses this action)
- Call the rocketCDN API to activate `PATCH /api/website/{rocketcdn_website_id}/` with:
```php
$data = [
			'is_active' => $is_active,// true or false
		];
```

### PR DESCRIPTION
**Pull request by Miraeld**
# Description

Fixes #8004 

This change automatically configures RocketCDN after a successful express checkout by detecting a URL parameter and enabling the CDN.

When users complete a purchase and are redirected back to their site with the `rocketcdn_checkout=true` parameter, WP Rocket will automatically:
1. Refresh user data to get fresh RocketCDN credentials
2. Store the token and enable CDN
3. Clean the URL and redirect

## Type of change

- [x] New feature (non-breaking change which adds functionality).

### What was tested

**Manual Testing:**
1. ✅ URL parameter `rocketcdn_checkout=true` triggers automatic configuration
2. ✅ User data is refreshed to get fresh cdn_token and cdn_url
3. ✅ Token is stored in `rocketcdn_user_token` option
4. ✅ CDN is enabled with the correct CDN URL
5. ✅ URL parameter is removed after processing via redirect
6. ✅ Guard clause prevents duplicate activation when token already exists
7. ✅ Permission check ensures only authorized users can trigger activation
8. ✅ Error handling for missing user data

**Integration Tests:**
1. ✅ `testShouldBailOutWhenParameterNotSet` - Verifies no processing when parameter is absent
2. ✅ `testShouldBailOutWhenUserLacksPermission` - Validates permission checks
3. ✅ `testShouldBailOutAndRedirectWhenTokenAlreadyExists` - Tests guard clause for existing tokens

### How to test

**Prerequisites:**
- WordPress site with WP Rocket installed
- Valid WP Rocket license

**Test Case 1: Normal Flow (First Activation)**
1. Ensure `rocketcdn_user_token` option does NOT exist (clean state)
2. Navigate to: `wp-admin/?rocketcdn_checkout=true`
3. **Expected:** 
   - User data cache is refreshed
   - `rocketcdn_user_token` option is created with the token
   - CDN is enabled in WP Rocket settings
   - Redirect to clean URL (parameter removed)
   - No errors in PHP error log

**Test Case 2: Guard Clause (Token Already Exists)**
1. Ensure `rocketcdn_user_token` option EXISTS (simulate previous activation)
2. Navigate to: `wp-admin/?rocketcdn_checkout=true`
3. **Expected:**
   - Method bails out early (guard clause)
   - Existing token remains unchanged
   - Redirect to clean URL

**Test Case 3: Permission Check**
1. Log in as user without `rocket_manage_options` capability
2. Navigate to: `wp-admin/?rocketcdn_checkout=true`
3. **Expected:**
   - Method returns early without processing
   - No configuration changes

**Test Case 4: Missing User Data**
1. Simulate scenario where user data API returns incomplete data (e.g., missing cdn_token)
2. Navigate to: `wp-admin/?rocketcdn_checkout=true`
3. **Expected:**
   - Method handles gracefully
   - No token stored
   - Redirect to clean URL
   - No fatal errors

### Affected Features & Quality Assurance Scope

**Affected Features:**
- RocketCDN express checkout flow
- User data caching and refresh logic
- RocketCDN CDN options enablement

**QA Testing Scope:**
- Express checkout flow end-to-end
- URL parameter detection on admin_init
- User data API integration
- Guard clauses for existing tokens
- Permission checks
- Error handling for various failure scenarios
- Integration with existing RocketCDN configuration flow

### Documentation

**Implementation:**

Added new functionality in `inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php`:

1. **Constructor Update**: Added `UserClient` dependency via DI
   - Allows fetching fresh user data with RocketCDN credentials

2. **Event Handler**: `handle_rocketcdn_checkout_parameter()` registered on `admin_init`
   - Detects `rocketcdn_checkout=true` URL parameter
   - Implements guard clause to prevent duplicate activation
   - Refreshes user data cache
   - Validates presence of required credentials (cdn_token, cdn_url)
   - Stores token and enables CDN using existing `cdn_options->enable()` method
   - Schedules subscription status check
   - Redirects to clean URL

3. **Helper Method**: `remove_query_parameter_and_redirect()`
   - Removes the query parameter from URL
   - Performs safe redirect to clean URL

**ServiceProvider Update** (`inc/Engine/CDN/RocketCDN/ServiceProvider.php`):
- Added `user_client` as dependency for `rocketcdn_data_manager_subscriber`

**PHPStan Baseline Update** (`phpstan-baseline.neon`):
- Updated counts for `get_option()` and `update_option()` usage in DataManagerSubscriber
- Maintains consistency with existing legacy code patterns

**Code flow:**
```
admin_init hook triggered
    ↓
handle_rocketcdn_checkout_parameter()
    ↓
Check for rocketcdn_checkout=true parameter
    ↓
Permission check: rocket_manage_options
    ↓
Guard clause: Check existing token
    ↓
Flush user data cache
    ↓
Fetch fresh user data (cdn_token, cdn_url)
    ↓
Validate credentials
    ↓
Store token in rocketcdn_user_token option
    ↓
Enable CDN with cdn_url (via existing cdn_options->enable())
    ↓
Schedule subscription check
    ↓
Remove query parameter & redirect
```

**Pattern used:**
- Follows Rocket Insights pattern for URL parameter handling (see `inc/Engine/Admin/RocketInsights/Subscriber.php::flush_license_cache()`)
- Uses dependency injection for UserClient (follows WP Rocket ServiceProvider pattern)
- Implements guard clauses for idempotency
- Reuses existing CDNOptionsManager::enable() method for consistent CDN enabling (KISS principle)

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Quality assurance

- [x] PHP Code Sniffer: All files pass (`composer phpcs`)
- [x] PHPStan: All files pass (`composer run-stan`)
- [x] Guard clauses implemented for edge cases
- [x] Error handling for missing user data
- [x] Permission checks enforced
- [x] Integration tests implemented and passing

### CODE REVIEW

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 1]
```diff
@@ -110,6 +121,76 @@ public function update_user_token() {
...
+	/**
+	 * Removes the query parameter and redirects to clean URL.
+	 *
+	 * @since 3.20.5
+	 *
+	 * @return void
+	 */
+	private function remove_query_parameter_and_redirect(): void {
+		$redirect_url = remove_query_arg( 'rocketcdn_checkout' );
+		wp_safe_redirect( $redirect_url );
```
[the remark this answers is not in the document]
Comment (Miraeld) [reply inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php:187]: Same pattern is used as RocketIsnights [here](https://github.com/wp-media/wp-rocket/blob/b07aca0277fdf529877028901fb4ba0c3fd07eb1/inc/Engine/Admin/RocketInsights/Subscriber.php#L468-L478) where we aren't using `exit()` or `die()`

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 2]
```diff
@@ -110,6 +121,76 @@ public function update_user_token() {
...
+	 * @since 3.20.5
+	 *
+	 * @return void
+	 */
+	public function handle_rocketcdn_checkout_parameter(): void {
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- Checking URL parameter for redirect, not processing form data.
+		if ( ! isset( $_GET['rocketcdn_checkout'] ) ) {
+			return;
+		}
+
```
[the remark this answers is not in the document]
Comment (Miraeld) [reply inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php:139]: Same pattern as RocketInsights [here](https://github.com/wp-media/wp-rocket/blob/b07aca0277fdf529877028901fb4ba0c3fd07eb1/inc/Engine/Admin/RocketInsights/Subscriber.php#L468-L478) we do not need a nonce.

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 3]
```diff
@@ -110,6 +121,76 @@ public function update_user_token() {
...
+		if ( false === $user_data || empty( $user_data->cdn_token ) || empty( $user_data->cdn_url ) || empty( $user_data->rocketcdn_website_id ) ) {
+			$this->remove_query_parameter_and_redirect();
+			return;
+		}
+
+		// Activate the subscription via RocketCDN API.
+		$activation_result = $this->api_client->activate_subscription(
+			sanitize_key( $user_data->cdn_token ),
+			(int) $user_data->rocketcdn_website_id
+		);
```
[the remark this answers is not in the document]
Comment (Miraeld) [reply inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: Added

File: inc/Engine/CDN/RocketCDN/APIClient.php [hunk 4]
```diff
@@ -116,6 +116,57 @@ public function get_pricing_data() {
...
+	/**
+	 * Activates the RocketCDN subscription via API.
+	 *
+	 * @since 3.20.5
+	 *
+	 * @param string $token      RocketCDN API token.
+	 * @param int    $website_id RocketCDN website ID.
+	 * @return true|WP_Error True on success, WP_Error on failure.
+	 */
+	public function activate_subscription( string $token, int $website_id ) {
```
Comment (mostafa-hisham) [inline inc/Engine/CDN/RocketCDN/APIClient.php:128]: API call looks ok.

**Comment by Miraeld** [comment]
Snippet to be used to change rocketcdn api to staging one:
``` php
add_filter('pre_http_request', function($pre, $args, $url) {
    static $doing_rocketcdn_request = false;
    
    // Prevent infinite recursion
    if ($doing_rocketcdn_request) {
        return $pre;
    }
    
    if (strpos($url, 'rocketcdn.me/api/') !== false) {
        // Replace production URL with staging URL
        $staging_url = str_replace('https://rocketcdn.me/api/', 'https://www-staging.rocketcdn.me/api/', $url);
        
        // Set flag to prevent recursion
        $doing_rocketcdn_request = true;
        $response = wp_remote_request($staging_url, $args);
        $doing_rocketcdn_request = false;
        
        return $response;
    }
    return $pre;
}, 9, 3);
```

File: inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php [hunk 5]
```diff
@@ -214,9 +214,10 @@ public function add_subscription_modal() {
 
 		$iframe_src = add_query_arg(
 			[
-				'website'  => home_url(),
-				'callback' => rest_url( 'wp-rocket/v1/rocketcdn/' ),
-				'source'   => 'plugin',
+				'website'       => home_url(),
+				'callback'      => rest_url( 'wp-rocket/v1/rocketcdn/' ),
+				'dashboard_url' => admin_url( 'options-general.php?page=' . WP_ROCKET_PLUGIN_SLUG . '&rocketcdn_checkout=true' ),
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php]: We shouldn't change the iframe src url at all.

**Review by wordpressfan — CHANGES_REQUESTED** [review]
I believe there is mis-understanding here, the dashboard_url should be added into the RocketCDN button url and I can't see this part's functionality implemented at all, so do we need another issue to handle the following part:
- Based on the response coming in user endpoint [here](https://github.com/wp-media/wp-rocket.me/issues/5499) if the `button` is there in the response, we will check if the `url` is there inside it we will add dashboard_url and then when the button is clicked we will open this full url in a new tab, but if it's not there we will open the iframe without adding the dashboard_url at all in this case

This should be independent of the UI issue because this is the missing piece to test E2E.

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 6]
```diff
@@ -110,6 +121,75 @@ public function update_user_token() {
...
+		if ( ! current_user_can( 'rocket_manage_options' ) ) {
+			return;
+		}
+
+		// Guard clause: bail if token already exists.
+		$current_token = get_option( 'rocketcdn_user_token' );
+		if ( ! empty( $current_token ) ) {
+			$this->remove_query_parameter_and_redirect();
+			return;
+		}
```
Comment (mostafa-hisham) [inline inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: @Miraeld @engahmeds3ed I think we should remove this part.
- Let's say the activation API call failed for any reason (website is inactive on rocketcdn side).
- The user will click on the express checkout link again.
- on wp-rocket.me side will check if the website is not active (inactive).
- show error message to the user, "please reactivate your website," with a button that redirects them to the dashboard.

How it looks on our end
Image

**Review by mostafa-hisham — CHANGES_REQUESTED** [review]

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php (hunk 6 above)
Comment (Miraeld) [reply inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: Removed :)

**Review by mostafa-hisham — COMMENTED** [review]
@Miraeld removed the condition that causes the activate issue.
@wordpressfan @wp-media/engineering-wpr-plugin-team 
Please continue code review.

**Comment by Mai-Saad** [comment]
@Miraeld Thanks for the PR. During the exploratory test

- [x]  can see that after purchase is done and user is redirected to WPR, the CDN banner is there while it shouldn't. can you please check that :pray:  (related thread https://group-onecom.slack.com/archives/C08F4LZB2QG/p1771492532500869)
- [x] When stripe account click get started and before redirect to website, banner loader is displayed https://group-onecom.slack.com/archives/C08F4LZB2QG/p1771491540328599?thread_ts=1771427653.990939&cid=C08F4LZB2QG => fixed on [01c0a8f](https://github.com/wp-media/wp-rocket/pull/8050/commits/01c0a8f2def6845d930803e3756d5793fcb0b25b)
- Page title at website after redirection isnot correct (to be fixed on staging saas) https://group-onecom.slack.com/archives/C08F4LZB2QG/p1771486230258189?thread_ts=1771427653.990939&cid=C08F4LZB2QG

File: inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php [hunk 7]
```diff
@@ -212,6 +221,15 @@ public function add_subscription_modal() {
 			return;
 		}
 
+		// Check if user data has button URL.
+		$button_url = '';
+		$user_data  = $this->user_client->get_user_data();
+
+		if ( false !== $user_data && isset( $user_data->rocketcdn->button->url ) && ! empty( $user_data->rocketcdn->button->url ) ) {
+			$dashboard_url = admin_url( 'options-general.php?page=' . WP_ROCKET_PLUGIN_SLUG . '&rocketcdn_checkout=true' );
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php]: Why not using `add_query_arg` to make it more readable.

File: inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php [hunk 8]
```diff
@@ -212,6 +221,15 @@ public function add_subscription_modal() {
...
+			$button_url    = add_query_arg( 'dashboard_url', rawurlencode( $dashboard_url ), $user_data->rocketcdn->button->url );
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php]: Why splitting this into two lines, do we really need this $dashboard_url?

File: inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php [hunk 9]
```diff
@@ -221,6 +239,9 @@ public function add_subscription_modal() {
 			'https://api.wp-rocket.me/cdn/iframe'
 		);
 		?>
+		<script type="text/javascript">
+			window.rocketcdnButtonUrl = <?php echo wp_json_encode( $button_url ); ?>;
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/AdminPageSubscriber.php]: What is the real need for wp_json_encode ?

```suggestion
			window.rocketcdnButtonUrl = <?php echo esc_js( $button_url ); ?>;
```

File: inc/Engine/CDN/RocketCDN/APIClient.php [hunk 10]
```diff
@@ -116,6 +116,57 @@ public function get_pricing_data() {
...
+		);
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$response_code = wp_remote_retrieve_response_code( $response );
+
+		if ( 200 !== $response_code ) {
+			return $this->get_wp_error(
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/APIClient.php:158]: Do we need to log all of those error message?
I'd say we can log in the method `get_wp_error` to cover all errors

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 11]
```diff
@@ -110,6 +121,71 @@ public function update_user_token() {
...
+
+		// Validate token length (must be 40 characters).
+		if ( 40 !== strlen( $user_data->rocketcdn->cdn_token ) ) {
+			$this->remove_query_parameter_and_redirect();
+			return;
+		}
+
+		// Activate the subscription via RocketCDN API.
+		$activation_result = $this->api_client->activate_subscription(
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: This is the part that We started a discussion about here:
https://group-onecom.slack.com/archives/C0ABFSS2PA6/p1771938334519989

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 12]
```diff
@@ -110,6 +121,71 @@ public function update_user_token() {
...
+			sanitize_key( $user_data->rocketcdn->cdn_token ),
+			(int) $user_data->rocketcdn->rocketcdn_website_id
+		);
+		update_option( 'rocketcdn_user_token', sanitize_key( $user_data->rocketcdn->cdn_token ) );
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: why not adding a new method inside `CDNOptionsManager` like `save_token` and then use it here like:
```
$this->cdn_options->save_token( sanitize_key( $user_data->rocketcdn->cdn_token ) );
```

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 13]
```diff
@@ -110,6 +121,71 @@ public function update_user_token() {
...
+
+		// Validate token length (must be 40 characters).
+		if ( 40 !== strlen( $user_data->rocketcdn->cdn_token ) ) {
+			$this->remove_query_parameter_and_redirect();
+			return;
+		}
+
+		// Activate the subscription via RocketCDN API.
+		$activation_result = $this->api_client->activate_subscription(
+			sanitize_key( $user_data->rocketcdn->cdn_token ),
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: we are sanitizing more than once, so it'd be perfect to make variables for `token`, and `website_id`

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php [hunk 14]
```diff
@@ -110,6 +121,71 @@ public function update_user_token() {
...
+			(int) $user_data->rocketcdn->rocketcdn_website_id
+		);
+		update_option( 'rocketcdn_user_token', sanitize_key( $user_data->rocketcdn->cdn_token ) );
+		$this->cdn_options->enable( esc_url_raw( $user_data->rocketcdn->cdn_url ) );
+
+		// Clear cached subscription status to reflect the newly activated subscription.
+		delete_transient( 'rocketcdn_status' );
```
Comment (wordpressfan) [inline inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php]: the above enable method already deletes this transient, no need to delete it again.

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (wordpressfan) [reply inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php:187]: I think it's more safe to exit here.

**Review by wordpressfan — CHANGES_REQUESTED** [review]

File: inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Miraeld) [reply inc/Engine/CDN/RocketCDN/DataManagerSubscriber.php:187]: If you exit, you can't have integration test..

**Comment by Miraeld** [comment]
@wordpressfan Thanks for the review! I've addressed all your feedback:

- Simplified the button URL logic in AdminPageSubscriber - removed the intermediate variable
- Changed `wp_json_encode()` to `esc_js()` for the JavaScript output
- Added a `save_token()` method to CDNOptionsManager to better encapsulate the token saving logic
- Refactored the checkout handler to avoid sanitizing values multiple times - now using variables
- Removed the duplicate `delete_transient()` call since it's already handled by the `enable()` method
- Kept the proper exit pattern with `rocket_get_constant()` for test compatibility

All tests are passing ✅

**Review by wordpressfan — APPROVED** [review]
Thanks @Miraeld 
We need to merge this one and create a GH issue for the remaining task that mentioned here:
https://group-onecom.slack.com/archives/C0ABFSS2PA6/p1771938334519989
